core/src/main/scala/org/apache/spark/shuffle/BlockStoreShuffleReader.scala

```
@@ -34,33 +39,58 @@ private[spark] class BlockStoreShuffleReader[K, C](
-      SparkEnv.get.conf.get(config.SHUFFLE_DETECT_CORRUPT),
-      readMetrics).toCompletionIterator
+    val wrappedStreams =
+      shuffleReadSupport.getPartitionReaders(new Iterable[ShuffleBlockInfo] {
+        override def iterator: Iterator[ShuffleBlockInfo] = {
+          /** Read the combined key-values for this reduce task */
+          mapOutputTracker.getMapSizesByExecutorId(handle.shuffleId, startPartition, endPartition)
+            .flatMap(blockManagerIdInfo => {
+              blockManagerIdInfo._2.map(
+                blockInfo => {
+                  val block = blockInfo._1.asInstanceOf[ShuffleBlockId]
+                  new ShuffleBlockInfo(block.shuffleId, block.mapId, block.reduceId, blockInfo._2)
+                }
+              )
+            })
+        }
+      }.asJava).iterator()
 
-    val serializerInstance = dep.serializer.newInstance()
+    val retryingWrappedStreams = new Iterator[InputStream] {
+      override def hasNext: Boolean = wrappedStreams.hasNext
 
-    // Create a key/value iterator for each stream
-    val recordIter = wrappedStreams.flatMap { case (blockId, wrappedStream) =>
+      override def next(): InputStream = {
+        var returnStream: InputStream = null
+        while (wrappedStreams.hasNext && returnStream == null) {
+          val nextStream = wrappedStreams.next()
+          val blockInfo = nextStream._1
+          val blockId = ShuffleBlockId(
+            blockInfo.getShuffleId,
+            blockInfo.getMapId,
+            blockInfo.getReduceId)
+          try {
+            returnStream = serializerManager.wrapStream(blockId, nextStream._2)
+            // Note: the asKeyValueIterator below wraps a key/value iterator inside of a
+            // NextIterator. The NextIterator makes sure that close() is called on the
+            // underlying InputStream when all records have been read.
+          } catch {
+            case e: IOException =>
```

Should this catch `Exception`?

---

So the original code has `IOException` because you want to catch only those, not `FetchFailedException` because that's an indicator from the `ShuffleBlockFetcherIterator` that it stopped retrying.

---

Ping here? Think the last push doesn't address this.

---

Let's try to add a comment allowing for the implementor to have some awareness, that this catches IOException specifically. Maybe to explain the intuition behind this logic. For example, why not catch all exceptions EXCEPT `FetchFailedException`?

---

Actually, I don't think I can add FetchFailedException because it's not part of the spark API... This is interesting because it also relates to @squito's concern about detecting failures that are used in the scheduler logic to mark all shuffle files on an executor as lost. Since other plugin implementations should be storing shuffle files remotely, it shouldn't mark one executor's shuffle files as lost, so we can perhaps make this a generic Exception and the plugins can throw whatever exceptions they want.

---

please see my rant below :)